### PR TITLE
Add ReactiveX deps, including JakeWharton's RxBinding and RxRelay

### PR DESCRIPTION
**Pull request by mgray88**
## 🚀 Description

Starting a draft PR to gather input on these dependencies, I will work on documentation in the meantime. There were a few things I wasn't sure about. Particularly the alias rules, I see some in txt files and some in the code, I wasn't sure what the recommended method was going forward. Also didn't really understand the removals and related generated things

Add ReactiveX dependencies:
- RxJava, RxJava2, RxJava3
- RxAndroid
- RxKotlin

Add Rx libraries from Jake Wharton:
- RxBinding3 and RxBinding4
- RxRelay2 and RxRelay3

## 📄 Motivation and Context

These are widely used libraries, many could probably benefit

## 🧪 How Has This Been Tested?

Ran tests. They passed!

## 📦 Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have read the [guidelines for the development process](https://jmfayard.github.io/refreshVersions/contributing/submitting-prs/dev-process/)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by brady-aiello** [comment]
- This should have an issue linked to it
- Looks like you need to update the autogenerated README entry:
https://github.com/jmfayard/refreshVersions/pull/478/files#diff-6529a4773c61668cf023627f4d039779841ae14af791dd2770f7c3811efdc55eR266

**Comment by mgray88** [comment]
Oops! Missed that requirement. Looks like there's an existing issue, I'll link it

I didn't really understand that autogenerated README entry. It's not really clear what I'm supposed to do with it?

File: plugins/dependencies/src/main/resources/refreshVersions-rules/reactivex-version-alias-rules.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+io.reactivex:*
+             ^
+
+io.reactivex.*:*
+               ^
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/refreshVersions-rules/reactivex-version-alias-rules.txt:5]: Left as is, one would get clashes if a project uses both RxJava 2 and RxJava3, which I believe is likely to happen since the idea behind using different maven coordinates for version 2 and 3 is precisely to allow this.
```suggestion
io.reactivex.*:*
             ^.^
```

File: plugins/dependencies/src/main/kotlin/dependencies/ReactiveX.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.gradle.kotlin.dsl.IsNotADependency
+
+@Incubating
+object ReactiveX : IsNotADependency {
+
+    val rxJava = DependencyNotation(group = "io.reactivex", name = "rxjava")
+
+    val rxAndroid = DependencyNotation(group = "io.reactivex", name = "rxandroid")
+
+    val rxKotlin = DependencyNotation(group = "io.reactivex", name = "rxkotlin")
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/ReactiveX.kt:15]: That's RxJava1, right? I'm pretty sure it's no longer updated, so I don't see much value in adding it into refreshVersions.

File: plugins/dependencies/src/main/resources/removals-revisions-history.md [hunk 3]
```diff
@@ -262,3 +262,11 @@ id:[androidx.car:car]
...
 ~~AndroidX.tracingKtx~~
 id:[androidx.tracing:tracing-ktx]
+
+## [WIP]
+
+~~ReactiveX.rxJava2~~
+// The 2.x version is end-of-life as of February 28, 2021. No further development, support, maintenance, PRs and updates will happen.
+// TODO: Migrate to io.reactivex.rxjava3
+moved:[io.reactivex.rxjava3:rxjava3]
+id:[io.reactivex.rxjava2:rxjava2]
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/removals-revisions-history.md:272]: This is intended for dependency notations that we are actually removing from refreshVersions. Here, you are adding it, so it should not be added here. Please, revert changes to this file.

File: plugins/dependencies/src/main/resources/removed-dependencies-versions-keys.txt [hunk 4]
```diff
@@ -51,3 +51,36 @@ androidx.legacy..legacy-support-v4=version.androidx.legacy
...
+com.jakewharton.rxbinding4..rxbinding=version.NO-RULE
+com.jakewharton.rxrelay2..rxrelay=version.NO-RULE
+com.jakewharton.rxrelay3..rxrelay=version.NO-RULE
+io.reactivex..rxjava=version.NO-RULE
+io.reactivex.rxjava2..rxandroid=version.NO-RULE
+io.reactivex.rxjava2..rxjava=version.NO-RULE
+io.reactivex.rxjava2..rxkotlin=version.NO-RULE
+io.reactivex.rxjava3..rxandroid=version.NO-RULE
+io.reactivex.rxjava3..rxjava=version.NO-RULE
+io.reactivex.rxjava3..rxkotlin=version.NO-RULE
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/removed-dependencies-versions-keys.txt:86]: You can remove all these lines, and re-run the tests, you'll see they are not added back. This was caused by a tests run order fact that wasn't taken into account, which is now fixed on the main branch.

File: plugins/dependencies/src/test/resources/dependencies-mapping-validated.txt [hunk 5]
```diff
@@ -747,6 +771,14 @@ io.pivotal.spring.cloud..spring-cloud-services-starter-config-client=Spring.spri
 io.pivotal.spring.cloud..spring-cloud-services-starter-service-registry=Spring.springCloud.serviceRegistry
 io.projectreactor..reactor-test=Spring.reactor.test
 io.projectreactor.kotlin..reactor-kotlin-extensions=Spring.reactor.kotlin
+io.reactivex..rxjava=ReactiveX.rxJava1
+io.reactivex..rxkotlin=ReactiveX.rxKotlin1
```
Comment (LouisCAD) [inline plugins/dependencies/src/test/resources/dependencies-mapping-validated.txt:775]: This looks inconsistent with what you have defined in the code (where there's no "1" suffix). Please, revert the changes on the files that are edited by the tests if you are editing the dependency notations in the Kotlin code.

Since we'll likely not include these dependency notations, it'll be obsolete of course, but at least, you know the process a bit better now.

File: plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt [hunk 6]
```diff
@@ -688,6 +712,14 @@ io.pivotal.spring.cloud..spring-cloud-services-starter-config-client=version.NO-
...
 io.projectreactor..reactor-test=version.NO-RULE
 io.projectreactor.kotlin..reactor-kotlin-extensions=version.NO-RULE
+io.reactivex..rxjava=version.rxjava
+io.reactivex..rxkotlin=version.rxkotlin
+io.reactivex.rxjava2..rxandroid=version.rxandroid
+io.reactivex.rxjava2..rxjava=version.rxjava
+io.reactivex.rxjava2..rxkotlin=version.rxkotlin
+io.reactivex.rxjava3..rxandroid=version.rxandroid
+io.reactivex.rxjava3..rxjava=version.rxjava
+io.reactivex.rxjava3..rxkotlin=version.rxkotlin
```
Comment (LouisCAD) [inline plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt:722]: As you can see, we have a problem here: dependencies that don't have their version synchronized have the same version key. Related to one of my other comments above.

**Review by LouisCAD — CHANGES_REQUESTED** [review]
Thanks for  the PR! There are some concerns, so I added comments in this review for you 🙂 

BTW, it might be easier for both of us to address these in a new PR, fresh from the current main branch, superseding this draft PR.

@brady-aiello It's a markdown file, but it's not a README 😄 
Anyway, the act of solely adding dependency notations should never touch this file, so… 🙃

File: plugins/dependencies/src/main/resources/refreshVersions-rules/reactivex-version-alias-rules.txt (hunk 1 above)
Comment (mgray88) [reply plugins/dependencies/src/main/resources/refreshVersions-rules/reactivex-version-alias-rules.txt:5]: Gotcha. I don't work with Android/Java/Kotlin as much anymore so I wasn't quite familiar with that. Just wanted to work on adding some deps for my team and the community

File: plugins/dependencies/src/main/kotlin/dependencies/ReactiveX.kt (hunk 2 above)
Comment (mgray88) [reply plugins/dependencies/src/main/kotlin/dependencies/ReactiveX.kt:15]: I initially didn't have it in there, but then I saw the RxJava1 adapter in Square.Retrofit2, so I figured I would add it. I agree though with you though so I'll remove it

File: plugins/dependencies/src/main/resources/removals-revisions-history.md (hunk 3 above)
Comment (mgray88) [reply plugins/dependencies/src/main/resources/removals-revisions-history.md:272]: Gotcha. I _thought_ maybe that was what it was for, but it wasn't completely clear. I added these thinking, hey if you're using RxJava2 you should update to RxJava3

File: plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt (hunk 6 above)
Comment (mgray88) [reply plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt:722]: Right. I think I actually did that on purpose with the idea that it would be simpler to move from RxJava2 to RxJava3 since the dependency name would be the same? But yea, I'll change it since I didn't realize people might want to import both at the same time.

**Comment by mgray88** [comment]
@LouisCAD Closing to reopen in a new PR after rebase and refactor :)

**Comment by LouisCAD** [comment]
I'd advise extracting a patch and apply select changes instead of rebasing, but maybe your plan will work just as well.

**Comment by LouisCAD** [comment]
Hello Mike, if you still plan on getting these added, please take the time to open the new PR as you told, or tell me you'll not do it, so I or someone else know they can do it. :)
